lihaoyi/mill
# Parallel task evaluation

This branch experiments with parallel task execution in mill.

This PR is an implementation of https://github.com/lihaoyi/mill/issues/32. Please check out and give feedback! If you need a more recent mill version, just ping me (@lefou) and I'll re-sync the branch.

_I try to update this description on the go, to reflect the current state. For details have a look at the commit messages. Everything is open for discussion, either here or on [Gitter](https://gitter.im/lihaoyi/mill). Just ping me._

## How to test

Checkout and build locally.

```
$ git clone -b par https://github.com/lefou/mill.git
Cloning into 'mill'...
$ cd mill
mill $ ./mill -i all __.publishLocal assembly
mill $ cp out/assembly/dest/mill ~/mill-par
```
The last line copies the mill executable to `~/mill-par`, but you can freely choose a different location. 

By default, mill uses as much build threads as available logical cpu cores. You can change it with `--jobs` or `-j` option. Use `-j 1` to disable parallel execution at all.

Please note, that the `all` target enforces serial processing to ensure that calls like `mill -i all clean __.compile` will still work. In most cases (e.g. when you don't include a `clean`) you might want to use the new `par` target, which runs the following targets in parallel. For example: `mill -i par __.compile __.testCached`.

If you found issues, please report!
If it works for you, please report, too!

## How to compare compile speeds

After a change of the mill version, various files need to be re-compiled (ammonite env, `build.sc` and possibly more). So, before you can compare build times with different mill versions you should run `mill -i version`, which ensures these things are done and wont foul your timings.

Also, if you want compare times side-by-side, you better clean your `out` directory with `rm -f out`.

For starters, if you just want to compare compile times, compare the outputs of:

```
rm -r out
time ~/mill-par -i __.compile
```

and 

```
rm -r out
time ~/mill-jar -i -j 1 __.compile
```

Don't expect wonders! The more cores you have, the better the speedup. Many Laptops throttle their cpu frequencies when they are gooing too hot. Higher cpu usage means more heat, so after all the build times may even increase because you are running in parallel. Look at the ratio between real and user time.

## Reporting bugs

If you think you found a bug, here are some infos that help me understanding and fixing it:

* Output of `mill-par --version`
* If possible a link to the public repository and steps to reproduce the error
* The build output, esp. the error message, also `out/evaluator.log` is interesting (or run mill with `-d`)
* Does the build succeed with `mill-par -j 1` ?
* Does the build succeed with mainline mill?

## Pull Request Details

### Task Evaluation

In this attempt, I use a Java `ExecutionService` for the job management, to avoid additional dependencies. It also allows some easy reasoning about job scheduling, as the control work and state handling is managed from the single self thread. Only Task execution happens in separate thread pool.

It uses a fixed thread pool with a configurable thread count. The thread pool is managed inside of Evaluator and created and shutdown for each evaluation phase (invokation of the `evaluate` def). That also means, that when running tasks which itself use an Evalutator instance, we end up with another thread pool. As an example `mill all clean __.compile` will run the `all` task in an outer `Evaluator` instance, which will block until an inner `Evaluator` instance completes the other tasks. Thus, I plan to redesign it to manage parallelism outside of Evaluator and increase the pool size for each new Evaluator instance, to gracefully adapt to the more or less blocking threads, that wait for the finish of the inner Evaluator instance.

Currently, we only parallelize whole terminal groups (which are most probably a collection of multiple tasks). All elements of a terminal group are by a high change depended on each other and also very fine grained, so that parallelization would not provide much gain; in contrary, it could produce more stress and overhead. But this is only an unverified assumption.

### Logging and output

Except for ticker message, I try to prefix all output that goes through the Logger with the current task name.

### Caveats

Tasks that get an `Evaluator`instance are potentially sensitive to parallel execution. For example the built-in `clean` task will most probably not work as expected when given as an argument to the built-in `all` task. This will be fixed once #502 is addressed.
